feat(spec): add CEP-15 Common Tool Schemas

ContextVM-org opened the pull request:
## Abstract

This CEP establishes a standard for defining and discovering common tool schemas in ContextVM. It enables interoperability by allowing multiple servers to implement standardized tool interfaces that clients can recognize and use consistently. Using MCP's `_meta` field, RFC 8785 for deterministic hashing, and CEP-6 announcements for discovery, this creates a marketplace where users can choose between multiple providers implementing the same standard tool interface.

Discussion and more info in the related issue

ContextVM-org committed:
feat(spec): add CEP-TBD Common Tool Schemas

hzrd149 commented on src/content/docs/spec/ceps/cep-21.md:
````diff
@@ -0,0 +1,415 @@
+}
+```
+
+**Tag Format**: `["D", "<schema-hash>", "<tool-name>"]` - Indicates this server defines a common tool schema.
````
What does it mean for a server to "define" a tool schema? if these schemas are deterministically hashed then it shouldn't matter which pubkey (server) defines them

ContextVM-org replied:
Hey! The original idea behind this distinction was to propose a native coordination layer, bootstrapping, and a signal for trust/reputation. However, after considering it carefully and keeping in mind adversarial scenarios as well, we agree that there is no need for this distinction or the authoritative mechanism. We have just committed a revision of the spec removing the 'D' tag.

ContextVM-org committed:
feat(spec): revise common tool schemas

Remove reference server pattern and simplify schema discovery to use
direct schema hashes without authoritative definers. Update announcement
tags to "I" for implemented schemas, eliminating "D" tags and reference
pubkey requirements. Enhance hash calculation notes and verification
flow for better interoperability.

ContextVM-org commented:
> It would be nice if the `I` tag were a NIP-73 reference.

Thanks you for the feedback, and I apologize for the delayed response @vitorpamplona, these days have been quite hectic. If I understood correctly, making the I tag a NIP-73 reference would mean declaring them in the NIP-73 document so that CVM servers can be referenced by the nips tags right? I cannot see any clear benefit, right now. Could you point us to why you think it would be beneficial for this specification? I'm not sure if treating the common schema hash as an external reference makes sense. It does make some sense, but currently cannot see any specific benefit, we can change our mind, so please share your rationale. Thanks

vitorpamplona commented:
NIP-73 serves as base to link external content into Nostr and make it part of the ecosystem. For instance, you can have NIP-22 comments and NIP-25 reactions about any NIP-73 identifier. It's just a common way of encoding different identifiers in one tag to make it indexable by relays. In that way, people can vote, follow, and discuss schemas directly.

ContextVM-org committed:
Merge branch 'master' into feat/common-schemas

ContextVM-org committed:
docs(spec): update cep-15 and cep-6

Revise CEP-15 to use NIP-73 compliant 'i' and 'k' tags instead of 'I' for schema discovery,
update examples and queries accordingly, and add NIP-73 as a dependency.
Update CEP-6 to reference CEP-15 for tools list announcements and add it as a dependency.

ContextVM-org committed:
Merge branch 'master' into feat/common-schemas

ContextVM-org commented:
Hey @abhayguptas, thanks for chiming in. Yes, you’re right, the current draft is a bit ambiguous about that. It makes sense to apply a narrow normalization before hashing that recursively removes these documentation oriented fields from the input and output schemas. I’ll update the draft now.

ContextVM-org committed:
docs(cep-15): add schema normalization for common tool schemas

Add schema normalization step to CEP-15 that removes JSON Schema `title` and `description` keywords before hash calculation, enabling compatible tool definitions from different providers to produce identical schema hashes while preserving documentation flexibility.

abhayguptas commented:
Following up on our discussion regarding the hashing boundaries in CEP-15, I wanted to document a few edge cases that could cause functionally identical tool schemas to produce divergent hashes.

### 1. Hash Inconsistency from Non-Functional Fields
Currently, the draft specifies removing `title` and `description`. However, JSON Schema allows many other documentation or metadata fields such as `examples`, `default`, `deprecated`, and `x-` vendor extensions. 

Because JCS (RFC 8785) does not delete fields, if Server A includes `"examples": ["Paris"]` and Server B omits it, JCS will preserve the `examples` key. This results in different hashes for functionally identical tools.

**Proposal:** We should expand the normalization step to explicitly strip `examples`, `default`, `deprecated`, and `x-*` fields before passing the object to JCS.

### 2. `$ref` Handling in Schemas
If a tool's `inputSchema` or `outputSchema` uses `$ref` to point to external or shared definitions, it is currently ambiguous how the hash is calculated.

If we hash the string literal `{"$ref": "#/defs/Location"}`, the hash becomes decoupled from the actual underlying schema logic. If the underlying definition changes, the hash stays the same. 

**Proposal:** We need to specify whether implementations MUST fully resolve/bundle `$ref` pointers *before* canonicalization and hashing.

### 3. Discovery Speed vs. Transport Bloat (CEP-35)
*Open architectural question:*
I was considering if we could use the new CEP-35 stateless discovery tags to let servers advertise their `schemaHash` values in the first-message exchange. This would allow clients to discover common tools instantly without needing an extra `tools/list` RPC call.

However, as discussed, this crosses the boundary between transport capabilities (CEP-35) and application state (`tools/list`). It could also severely bloat the session envelope if a server implements many tools. Documenting this here to confirm that keeping tool hashes strictly attached to the `tools/list` response is the preferred path.

ContextVM-org committed:
docs(cep-15): enhance schema normalization with $ref handling and expanded field exclusions

Adds hash payload boundary definition, requires self-contained schema representation for $ref, expands normalization to exclude examples, default, deprecated, readOnly, writeOnly, and x-* vendor fields. Updates client conformance verification steps and implementation guidance.

ContextVM-org commented:
Thanks for the review. We updated the draft to address these edge cases.

- **Non-functional fields:** the hash boundary is now explicit: only `name`, normalized `inputSchema`, and normalized `outputSchema` participate in the hash. Within schemas, CEP-15 now strips defined annotation/metadata fields and `x-*` vendor extensions before canonicalization.
- **`$ref` handling:** the spec now requires hashing a self-contained schema representation rather than unresolved `$ref` strings, and it forbids live remote resolution during hash calculation.
- **Discovery scope:** we kept schema-hash discovery on the existing CEP-15 surfaces rather than moving it into CEP-35. The draft now clarifies that direct ContextVM [`tools/list`](src/content/docs/spec/ceps/cep-15.md:377) responses and CEP-6 tools announcements reuse the same underlying payload, so common-schema references should be carried consistently across both, including event-level `i` / `k` tags.

Let me know what do you think @abhayguptas

abhayguptas commented:
Thanks for the quick update! I just pulled and reviewed the commit. The expanded normalization rules (stripping examples, default, x-*, etc.) and the self-contained $ref requirement look perfect and completely solve the edge cases.

I also really like the approach for discovery. Attaching the i and k tags directly to the tools/list response event is a great middle-ground—it keeps the CEP-35 transport layer clean from bloat while still giving clients fast event-level discovery. This looks rock solid to me! 🙂

ContextVM-org commented:
Great! Would you like to update your pr with this changes covering the edge cases so we can move forward?

abhayguptas commented:
Hey! I've updated the PRs based on our recent review.

ContextVM-org committed:
Merge branch 'master' into feat/common-schemas
